Modify that web action in the bound package can be accessed.

upgle opened the pull request:
<!--- Provide a concise summary of your changes in the Title -->

## Description
<!--- Provide a detailed description of your changes. -->
<!--- Include details of what problem you are solving and how your changes are tested. -->

Recently, many users want to share their web actions with the package.
but, the web action in the bound package can't be accessed in the browser. 

Because when it is invoked in the browser, it doesn't resolve an action to find original fully qualified action name.  so, I added the action name resolver.

#### HOW TO TEST.
bind a package that contains web actions and access web action which is in the bound package using your browser, you may see errors like the image below.

![image](https://user-images.githubusercontent.com/5635513/42751279-37e45712-8925-11e8-9555-51e7b689206b.png)

## Fixed
Now, the bound web action can be accessed in the browser.

![image](https://user-images.githubusercontent.com/5635513/42752053-b4b175b6-8927-11e8-92e2-a24c52a7b8ef.png)


## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [x] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [x] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

rabbah commented:
Thanks for trying this. There are some issues to be aware of packaged actions as web actions via bindings. The lack of support isn’t a bug (it was intentional at the time). This has come up before so it’s worthwhile to try and enable this. 

1. If an action isn’t already a web action in the package, it can’t be accessed as a webaction through the biding. That is a subject that binds the action can’t make it a web action (unless they already own it). So the creator/shares of the package will have to decide which of the actions to make available as web actions. 

2. The entitlement check for a web action today is straightforward: namespace must own both and so implicitly entitled to them. With a binding you need to check if the binding continues to have rights to the bound package.

3. And of course you’ll need to add tests for the change/new feature.

upgle commented:
@rabbah 
Thank you for your feedback and sorry that I thought it was a bug. 
I will add some test codes including the rights check and suggest it again.

upgle commented:
@rabbah 

I modified it to check the entitlement and added tests.
Now, It returns the authentication error if a web action has no entitlement.
```
{
  "error": "The supplied authentication is not authorized to access 'whisk.system/share'.",
  "code": "NmSIY4dJanMT7YS3MPwFoOPL4Np72g5B"
}
```

Could you please remove WIP label and review again?

upgle commented:
I fixed a conflict. please review it.

rabbah commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -700,9 +713,11 @@ trait WhiskWebActionsApi extends Directives with ValidateRequestSize with PostAc
-      case _: ArtifactStoreException | DeserializationException(_, _, _) =>
-        Future.failed(RejectRequest(NotFound))
+    resolveAction(actionName) flatMap { resolveAction =>
+      getAction(resolveAction) recoverWith {
```
i dont believe this is right - the resolveAction method returns just the name of the action in the base package, and so getting that action means you'll miss the bound parameters. I think you want to use resolveActionAndMergeParameters instead.

upgle replied:
Oh... I understand I'll modify it soon.

rabbah commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -743,6 +758,16 @@ trait WhiskWebActionsApi extends Directives with ValidateRequestSize with PostAc
+    implicit transid: TransactionId): Future[Unit] = {
+    val resource =
+      Resource(action.namespace.root.toPath, Collection(Collection.PACKAGES), Some(action.namespace.last.toString))
+    entitlementProvider.check(identity, Privilege.READ, resource)
```
this check also checks the throttles - which means there's now a redundant check for throttles.
the code (before this pr), did not check entitlement through the provider because it can implicitly infer the owner (there's no indirection through a binding) so it only checked throttles.

upgle replied:
Thank you for your kind review. I've removed existing entitlementProvider.checkThrottles(), then modified to use only entitlementProvider.check ().

upgle commented:
@rabbah Maybe it's done. could you please review it again? 😅

upgle commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -552,8 +552,11 @@ trait WhiskWebActionsApi extends Directives with ValidateRequestSize with PostAc
+          checkEntitlement(actionOwnerIdentity, a) flatMap { _ =>
+            pkgLookup(a.namespace.toFullyQualifiedEntityName) map { pkg =>
+              (a.inherit(pkg.parameters))
+            }
```
@rabbah 
yes, I agree that it looks complicated. so I've changed some codes. (removed `pkgLookup` method which is not necessary anymore)

1. action in a default package no need to resolve, just check entitlement and if it isexported.
2. action in a bound package(new) is handled by `resolveActionAndMergeParameters()`, it would resolve action and merge parameters. 
2. check whether action is in a proper package -> It check whether an action is exported by `confirmExportedAction()` method
3. and all types of action is checked for entitlement and throttle by `checkEntitlement()` method

please check changed code here

```scala
private def verifyWebAction(actionName: FullyQualifiedEntityName, authenticated: Boolean)(
    implicit transid: TransactionId) = {

    // lookup the identity for the action namespace
    identityLookup(actionName.path.root) flatMap { actionOwnerIdentity =>
      confirmExportedAction(actionLookup(actionName), authenticated) flatMap { a =>
        checkEntitlement(actionOwnerIdentity, a) map { _ => (actionOwnerIdentity, a)}
      }
    }
  }
```

https://github.com/apache/incubator-openwhisk/pull/3880/files#diff-a356e6b010b8aeb5ea7b6f7fec395038R535

junoyoon commented:
@upgle is it done?

rabbah commented:
The last commit looks a lot better now - I'll review more thoroughly.

upgle commented:
@junoyoon no, i'm waiting for review.

@rabbah ok, tell me if I missed something.

rabbah commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -440,21 +441,14 @@ trait WhiskWebActionsApi extends Directives with ValidateRequestSize with PostAc
+   * This method is factored out to allow mock testing
    */
-  protected def getAction(actionName: FullyQualifiedEntityName)(
+  protected def resolveActionAndMergeParameters(actionName: FullyQualifiedEntityName)(
```
We should add a comment that while parameters are merged, annotations are not. This is important because the web annotation is on an action not an entire package. We should separately add a test for this invariant.

rabbah replied:
need to update the docs here: https://github.com/apache/incubator-openwhisk/blob/master/docs/webactions.md#additional-features the precedence order would now also include package binding parameters.

upgle replied:
ok I'll update the docs soon.

rabbah replied:
thank you - it might be worth adding a section in the docs about web actions for packages and bindings (taking in some of my comments below). what do you think?

rabbah commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -537,27 +531,15 @@ trait WhiskWebActionsApi extends Directives with ValidateRequestSize with PostAc
-            (a.inherit(pkg.parameters))
-          }
+    // lookup the identity for the action namespace
+    identityLookup(actionName.path.root) flatMap { actionOwnerIdentity =>
```
here the identity is tied to how the action is invoked: if the path is relative to the package, it's the owner of the package. But if it's via the binding, the identity is the owner binding. The two identities are not the same. It might also not make sense to invoke an action in a public package since it might require a binding to work correctly.

rabbah replied:
another aspect to document: a web action in a public package cannot be disabled so any (private) binding of the package will have the web action exposed.

rabbah commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -684,33 +666,14 @@ trait WhiskWebActionsApi extends Directives with ValidateRequestSize with PostAc
-        // fail the request with BadRequest so as not to leak information about the existence
-        // of packages that are otherwise private
-        logging.debug(this, s"package which does not exist")
-        Future.failed(RejectRequest(NotFound))
```
we should add some tests for `resolveActionAndMergeParameters` to preserve the behavior represented by these request rejections (missing package or malformed package).

rabbah commented on tests/src/test/scala/whisk/core/controller/test/WebActionsApiTests.scala:
```diff
@@ -233,7 +238,7 @@ trait WebActionsApiBaseTests extends ControllerTestCommon with BeforeAndAfterEac
         Future.successful(theAction)
       } else {
-        getPackage(actionName.path.toFullyQualifiedEntityName) map (_ => theAction)
+        getPackage(actionName.path.toFullyQualifiedEntityName) map (pkg => theAction.inherit(pkg.parameters))
```
there's no test for actually using a package binding here (if there is, it would fail since the parameter inheritance does not include the binding).

ningyougang commented:
@upgle ,can you rebase?

upgle commented:
@ningyougang thank you, i'll rebase it.

rabbah commented:
I rebased this and resolved the conflict.

upgle commented on docs/webactions.md:
````diff
@@ -485,6 +485,12 @@ $ curl https://${APIHOST}/api/v1/web/guest/default/custom-options.http -kvX OPTI
 ```
 
+## Sharing web actions
+
````
@rabbah i've checked what you changed. thank you very much!

upgle commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -697,22 +672,15 @@ trait WhiskWebActionsApi
-        Future.failed(RejectRequest(NotFound))
-      case _: NoSuchElementException =>
-        logging.debug(this, s"'$pkg' is a binding")
-        Future.failed(RejectRequest(NotFound))
```
@rabbah 
test cases for 2, 3 have been added. https://github.com/apache/incubator-openwhisk/pull/3880/commits/e970d3a54a207fb60c41b7c56b2d0e802434428c

rabbah commented:
Thanks - will check this out today or tomorrow.

upgle commented:
@rabbah ok, thank you

rabbah commented on tests/src/test/scala/whisk/core/controller/test/WebActionsApiTests.scala:
```diff
@@ -126,36 +122,47 @@ trait WebActionsApiBaseTests extends ControllerTestCommon with BeforeAndAfterEac
+  override def beforeAll() = {
+    implicit val tid = transid()
+
+    put(entityStore, stubPackage, garbageCollect = false)
```
this is why the stubs arent collected: garbageCollect is set to false.

upgle replied:
yes, it seems other test case remove all document after tests.

https://github.com/apache/incubator-openwhisk/blob/master/tests/src/test/scala/whisk/core/controller/test/ControllerTestCommon.scala#L203

rabbah commented:
I fixed the test fixture cleanup and also tweaked one test.

upgle commented:
@rabbah test is passed. thank you for your support 👍

rabbah commented:
Thank you @upgle for your contribution.
